tools/server/webui/src/utils/llama-vscode.ts

```diff
@@ -41,7 +44,7 @@ export const useVSCodeContext = (textarea: ChatTextareaApi) => {
 
     window.addEventListener('message', handleMessage);
     return () => window.removeEventListener('message', handleMessage);
-  }, [textarea]);
+  }, [textarea, extraContext]);
```

Comment:
Should we make the "extra" context go first? The reason is that cache reuse will be utilized better - different requests with the same file will be able to reuse the context even if the textarea input is different.

Reply:
Both `textarea` and `extraContext` object will be re-created on every render, so for now I think the order is not very important.

While the members inside the object may not changed, the object is created on every render which make this `useEffect` see it as different values (yike, javascript thing 😂 ) . Ofc a better way is to listen to the exact attribute, like `textarea.refOnSubmit`. But I think we can investigate this later on. Normally, the cost of these operations are minor.

Reply:
Ah btw the `useEffect` will be triggered if one of the value in the listener array change, so the order is actually not important

Reply:
I see - my point is different though. I want to make the prompt to look like this:

```
[contents of files / extra context]
[contents of textarea input]
```

I think with the current implementation it is the other way around - the textarea input is placed before the contents of the files.

This way, I can pass a large file in one conversation and ask question 1. Then I can start a new conversation with the same large file and ask question 2. With prefix caching, the second conversation will process very fast. Similar effects will be observed with cache reuse.

Reply:
Oh ok I see your point, yes the user message is supposed to be before the extra context. This is decided by `normalizeMsgsForAPI` in `misc.ts` which converts the internal message representation to OAI API.

Fixing this now, thanks for noticing!

Reply:
Fixed in https://github.com/ggml-org/llama.cpp/pull/13365/commits/9a24d3f07b1361fd69d5098fe0e934f3ecbd26b7